docs/glossary.md

```diff
@@ -0,0 +1,204 @@
+given [operation](#operation). Note that this is quite different from Git's
+[HEAD](https://git-scm.com/book/en/v2/Git-Internals-Git-References#ref_the_ref).
+
+## Operation
+
+A snapshot of the [visible commits](#visible-commits) and [branches](#branches)
+at a given point in time (technically a [view object](#view)), together with
+some metadata. The metadata includes the username, hostname, timestamps, and
+pointers to the operation's parents.
+
+## Operation log
+
+The operation log is the
+[DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) formed by
+[operation](#operation) objects, much in the same way that commits form a DAG,
+which is sometimes called the "commit history". 
+
+## Repository
+
+Basically everything under `.jj/`, i.e. the full set of [operations](#operation)
+and [commits](#commit).
+
+## Remote
+
+TODO
+
+## Revision
+
+A synonym for [Commit](#commit).
+
+## Revset
+
+Jujutsu supports a functional language for selecting a set of revisions.
+Expressions in this language are called "revsets". See [here](revsets.md) for
+details. We also often use the term "revset" for the set of revisions selected
+by a revset.
+
+## Rewrite
+
+To "rewrite" a commit means to create a new version of that commit with
+different contents, metadata (including parent pointers), or both. Rewriting a
+commit results in a new commit, and thus a new [commit ID](#commit-id), but the
+[change ID](#change-id) generally remains the same.
+
+## Root commit
+
+The root commit is a virtual commit at the root of every repository. It has a
+commit ID consisting of all '0's (`00000000...`) and a change ID consisting of
+all 'z's (`zzzzzzzz...`). It can be referred to in [revsets](#revset) by the
+special name `root`. Note that our definition of "root commit" is different from
+Git's; Git's "root commits" are the first commit(s) in the repository, i.e. the
+commits `jj log -r root+` will show.
+
+## Tree
+
+A tree object represents a snapshot of a directory in the repository. Tree
+objects are defined recursively; each tree object only has the files and
+directories contained directly in the directory it represents.
+
+## Visible commits
+
+Visible commits are the commits you see in `jj log -r 'all()'`. They are the
+commits that are reachable from an anonymous head in the [view](#view).
+Ancestors of a visible commit are implicitly visible.
+
+## View
+
+A view is a snapshot of branches and their targets, anonymous heads,
+and working-copy commits. The anonymous heads define which commits
+are [visible](#visible-commits).
+
+A view object is similar to a [tree](#tree) object in that it represents a
+snapshot without history, and an [operation](#operation) object is similar to a
+[commit](#commit) object in that it adds metadata and history.
+
+## Workspace
+
+A workspace is a [working copy](#working-copy) and an
+associated [repository](#repository). There can be multiple workspaces for a
```

Comment:
Maybe "a pointer to the associated repository"?

Reply:
I'm a little worried that would make the reader think that the repository always lives somewhere else. Isn't the second sentence in the paragraph enough to explain that they can be shared?

Reply:
I find this mildly confusing (if a workspace contains a repository, how could there be multiple workspaces per repository), but this is a minor issue. If we can't think of a clearly better phrasing, let's leave it as is for now.

Reply:
I added a bit more detail here. PTAL. Btw, this is also related to the definition of "repository". In the source code and in the structure under `.jj/`, the repo (`.jj/repo/`) is separate from the working copy and state (`.jj/working_copy/`).

Reply:
Looks good, I think. Thanks!

> In the source code and in the structure under .jj/, the repo (.jj/repo/) is separate from the working copy and state (.jj/working_copy/).

Yes, I think that's a good distinction. The definition of a repository should really be "everything under `.jj/repo`" (I'm not proposing we change it right now in this doc, though -- if we do that and nothing else, it will probably only look more confusing). I think and hope people won't be confused by the explanation as it is now, but we can try to make it more precise later if we don't figure out a perfectly precise wording right now.